### ISSUE TITLE
Get error with PadV2 unsupported

### ISSUE THREAD

**Issue by fullm00n1** [ARM-software/armnn#603]
I get the following error when trying to load a tflite graph that was generated from converting an ONNX model.  It would be helpful if someone could help point me in the right direction:

Error: Failed to parse operator #23 within subgraph #0 error: Operator not supported. subgraph:0 operator:23 opcode_index:6 opcode:60 / PADV2  at function ParseUnsupportedOperator [armnn-devenv/arm
nn/src/armnnTfLiteParser/TfLiteParser.cpp:923]

I am running ArmNN 21.11 and Computer Library 21.11

**Comment by MikeJKelly** [comment]
Hi @fullm00n1 

I've just looked and we don't currently support PADV2 on the TFLiteParser, we do support PAD so it should be easy to add support for PADV2. I'll see if I can get a patch up today. 

ArmNN can only handle PADV2's with a single constant_value but as you're coming from ONNX that shouldn't be a problem.

Best regards,
Mike

**Comment by fullm00n1** [comment]
How does one tell if my model only has PADV2's with a single constant value?  We believe the PadV2 instances are due to the onnx-tf converter functionality in the Tensorflow tools but aren't aware of how to avoid the PadV2 as part of the conversion process.

**Comment by MikeJKelly** [comment]
Hi @fullm00n1 

There is a review with PADV2 support at:

https://review.mlplatform.org/c/ml/armnn/+/6908

Can you see if it solves your problem?

From my reading of the ONNX documentation at https://github.com/onnx/onnx/blob/master/docs/Operators.md#Pad ONNX only supports PAD with a single constant value. If there are multiple constant_values present in your model then the TFLite parser will throw an exception: "Multiple padding values are not supported in PADV2". 

Best regards,
Mike

**Comment by fullm00n1** [comment]
I integrated the changes in and ran the application I have that uses the armnn libraries.  I first verified the application and parser with a tflite model that doesn't have the PadV2.  That model ran as expected.  I then ran the newer model with the PadV2 layers.  I hit an assertion in the AddPadLayer of the TfLiteParser.  It looks to be something in the pad descriptor based on the call stack and the compilers limited symbols reported.  

My data type for all layers in the model is Float32.

 The specific assertion was a malloc call:

malloc.c:2406: sysmalloc: Assertion `(old_top == initial_top (av) && old_size == 0) || ((unsigned long) (old_size) >= MINSIZE && prev_inuse (old_top) && ((unsigned long) old_end & (pagesize - 1)) == 0)' failed.

**Comment by MikeJKelly** [comment]
That might be a problem trying to copy the data for the constant_value or the paddings from the TFLite model. Can you run your application under valgrind? That might give us a better idea of what's going wrong. https://valgrind.org/

Otherwise can you tell us a little more about your PADV2 layer? You can visualise your TFLIte model using Netron: https://netron.app/

And then look for the PADV2 nodes in the Graph, select them and confirm that it has values for the pad constant value and the paddings?

**Comment by fullm00n1** [comment]
Running valgrind might be hard as we are running a Xilinx Zynq Ultrascale+ (custom ARM64 environment).  Also, netron isn't a real option as our network is proprietary and we have restrictions on what tools / apps we can use with our graph/model.   I have attached a sanitized report from tensorflow of the graph report post ONNX to TF translation with all the Pad and PadV2 layer info from our model.  Does this help?

![tensorflow_report](https://user-images.githubusercontent.com/97263872/149005631-1218979f-f344-45d5-a12d-bca824329d47.JPG)

**Comment by fullm00n1** [comment]
I tried compiling and running the valgrind tool but it chokes because of my custom linux build for the custom platform we are running on.

**Comment by fullm00n1** [comment]
I can do some old school print statements with count in the ParsePad function if that would help?  Not exactly sure what to instrument up but given that it appears to break at the m_Network->AddPadLayer call.  Is there something specific to instrument on the inputs to that call?

**Comment by MikeJKelly** [comment]
It looks like the const_values are missing from the PADV2 (Tensor 6 PadV2/constant_values has a Shape of []). I've updated https://review.mlplatform.org/c/ml/armnn/+/6908 to hopefully better handle this. Could you please check if it prevents the issue with the failed assertion?

**Comment by fullm00n1** [comment]
I incorporated the changes you requested and tried to run again.   It gives the same malloc error as the previous code you posted.

I then added some traditional printing statements to try and find where memory issue is. 

Within the TFLiteParser.cpp, I added the following line of code prior to the "m_Network->AddPadLayer":

std::cout << layerName.c_str() << " -- " << desc.m_PadValue << " -- " << desc.m_PadList.size() << "\n" << std::flush;

I get the following output before it hits the assertion:

Pad:0:6 -- 0 -- 4
Pad:0:16 -- 0 -- 4
PadV2:0:23 -- -3.40282e+38 -- 4

The PadV2 is where is hits the assertion.

I then instrumented up NetworkImpl::AddPadLayer and Graph::AddLayer(Args&&... args)

The asserting occurs within the following call within Graph::AddLayer

LayerT* const layer = new LayerInGraph<LayerT>(*this, std::forward<Args>(args)...);

It appears to fail in LayerInGraphBase(Graph& graph, Iterator insertBefore, Args&&... args)
        : LayerT(std::forward<Args>(args)...), m_Graph(&graph).

Possibly in either Layer constructor or the Graph copy.   

Ideas on where to go with this?

**Comment by fullm00n1** [comment]
A little more info based on my print statements

I updated the PadLayer constructor to add a print statement:

                                              PadLayer::PadLayer(const armnn::PadDescriptor& param, const char* name)
                                                  : LayerWithParameters(1, 1, LayerType::Pad, param, name)
                                              {
                                              std::cout << "At entry to PadLayer::PadLayer function\n" << std::flush;
                                              }

I also updated the LayerWithParameters constructor with a print statement:

                                        LayerWithParameters(unsigned int numInputSlots,
                                                            unsigned int numOutputSlots,
                                                            LayerType type,
                                                            const Parameters& param,
                                                            const char* name)
                                            : Layer(numInputSlots, numOutputSlots, type, name)
                                            , m_Param(param)
                                        {
	                                    std::cout << "At entry to LayerWithParameters constructor function\n" << std::flush;
                                        }

On the PadV2 AddLayer call, it appears to be a problem with arguments to LayerWithParameters as I don't get the print statement.  That probably means there is something going on in one of the copy constructors.

I also put print statement in the Layer constructor like below and don't get any of those print outs:

                                      Layer::Layer(unsigned int numInputSlots,
                                                   unsigned int numOutputSlots,
                                                   LayerType type,
                                                   DataLayout layout,
                                                   const char* name)
                                      : m_OutputHandlers(numOutputSlots)
                                      , m_ShapeInferenceMethod(ShapeInferenceMethod::ValidateOnly)
                                      , m_LayerName(name ? name : "")
                                      , m_Type(type)
                                      , m_BackendId()
                                      , m_BackendHint(EmptyOptional())
                                      , m_Guid(profiling::ProfilingService::GetNextGuid())
                                      {
                                      std::cout << "At entry to Layer::Layer function\n" << std::flush;
                                          IgnoreUnused(layout);
                                          m_InputSlots.reserve(numInputSlots);
                                          for (unsigned int i = 0; i < numInputSlots; ++i)
                                          {
                                              m_InputSlots.emplace_back(*this, i);
                                          }
                                      
                                          m_OutputSlots.reserve(numOutputSlots);
                                          for (unsigned int i = 0; i < numOutputSlots; ++i)
                                          {
                                              m_OutputSlots.emplace_back(*this, m_OutputHandlers[i]);
                                          }
                                      std::cout << "At exit to Layer::Layer function\n" << std::flush;
                                      }

So the LayerInGraph method gets called in Graph::AddLayer, but the subsequent constructors fail somewhere before getting to any of my print statements in the PadLayer, LayerWithParameters, and Layer class constructors.

**Comment by fullm00n1** [comment]
A little more info from tensorflow on the tflite graph.   The PadV2 with op index of 23 (the PadV2 that is failing above), states that it has the following inputs:

85 onnx_tf_prefix_fire3_concat FLOAT32 [1, 128, 56, 56][-1, 128, 56, 56]
57 PadV2/paddings INT64 [4, 2]None
61 PadV2/constant_values FLOAT32 []None

And its output is:

86 PadV2 FLOAT32 [1, 128, 57, 57][-1, 128, 57, 57]

I am more a software guy and not a Tensorflow or Neural Network guy so I am a little lost at the Pad vs PadV2 operator and what this specifically means but thought it might help inform why the Parsing of the PadV2 operator is problematic from the TfLite model.

**Comment by fullm00n1** [comment]
![image](https://user-images.githubusercontent.com/97263872/149383580-71cbe1d5-307f-49af-b182-f28cfbf0968f.png)

**Comment by MikeJKelly** [comment]
It looks like the problem is with paddings data type (Int64). I'm not sure why that's int64 rather than int32 but looking at https://www.tensorflow.org/api_docs/python/tf/pad TF _seems_ to expect "paddings to be A Tensor of type int32.", in python at least. 

In our TFLite parser we end up trying to memcpy the uint64 values there into a uint32 vector which causes a crash. I've updated the review with a test where the paddings are int64 and it seems to be working now. Can you try it with your model please? 

https://review.mlplatform.org/c/ml/armnn/+/6908

**Comment by fullm00n1** [comment]
Mike,

Again, thanks for your help on this issue.

The TfLiteParser fix appears to have solved the memory allocation issue.  Now, my call to CreateNetworkFromBinaryFile executes successfully.   However, I was not able to run inference using the model because my call to "armnn::Optimize" fails.   It fails in Layer::VerifyShapeInferenceType.   I placed a print statement at the beginning of that method to determine which layer is the issue.  This was the output:

![image](https://user-images.githubusercontent.com/97263872/149578194-28f7831d-a6bc-4aa6-b93c-4b91d19fe52a.png)

I am not sure why the Sub op has a problem.  To review my conversion process, I first take our ONNX model and run onnx-tf on it to convert it to a Tensorflow Pb saved model.  I then import the saved model in Tensorflow and use the converter within Tensorflow to convert it to a tensorflow lite model.   That is what I am using with the ArmNN library when I call CreateNetworkFromBinaryFile.

So at this point, I am still not able to run the model for inference and verify the PadV2 operates.  Suggestions on what might help diagnose the problem during validation?

There is only one Sub op in the model.  In the Tensorflow model, the Sub op shows up in Tensorboard as shown below:

![image](https://user-images.githubusercontent.com/97263872/149578919-1841653e-4f3d-45b7-acd1-9fdf1c2414c0.png)

![image](https://user-images.githubusercontent.com/97263872/149578996-cdfa3b7c-c6ca-458d-8dde-3552086985f6.png)

![image](https://user-images.githubusercontent.com/97263872/149579034-01e9b882-9d13-4988-bb26-d35bfcb0b795.png)

Running the tensorflow.lite.tools.visualize on the tensorflow lite mode, I get this info regarding the Sub Op in the Tensorflow lite converted model:

![image](https://user-images.githubusercontent.com/97263872/149580826-da42dd9b-8afa-4cd6-a4fb-24c2cd3803e8.png)

I feel like we are getting closer.

**Comment by MikeJKelly** [comment]
Ah, if your ShapeInferenceMethod is set to ShapeInferenceMethod::InferAndValidate instead of ShapeInferenceMethod::ValidateOnly then ArmNN will attempt to figure out valid shapes when the shapes are dynamic. 

Are you running your model through ExecuteNetwork?  If you are then you can run the program with the --infer-output-shape command line switch to use  ShapeInferenceMethod::InferAndValidate 

If you're running it from your own code then you'll have to switch your ShapeInferenceMethod to InferAndValidate while you're optimizing your networks. Something like this:

```
OptimizerOptions optOptions;
optOptions.m_shapeInferenceMethod = armnn::ShapeInferenceMethod::InferAndValidate;
IOptimizedNetworkPtr optNet = Optimize(*net, backends, runtime->GetDeviceSpec(), optOptions);
```

**Comment by fullm00n1** [comment]
Mike,
   We are running the API in our own code.   I tried to add the OptimizerOptions to the Optimize called but got the same error:

![image](https://user-images.githubusercontent.com/97263872/149813670-9207ced2-af2c-4035-a20f-58755b73a67f.png)

Here is a snippet of code from our application that I am using to create and load the network.  Am I missing something?

          // Import the TensorFlowLite model.
          using IParser = armnnTfLiteParser::ITfLiteParser;
          auto armnnparser(IParser::Create());
          armnn::INetworkPtr network = armnnparser->CreateNetworkFromBinaryFile(programOptions.modelPath.c_str());
          
          // Find the binding points for the input and output nodes
          using BindingPointInfo = armnnTfLiteParser::BindingPointInfo;
          const std::vector<BindingPointInfo> inputBindings  = { armnnparser->GetNetworkInputBindingInfo(0, inputName) };
          const std::vector<BindingPointInfo> outputBindings = { armnnparser->GetNetworkOutputBindingInfo(0, outputName) };
          
          // ------------------------------------------------------------------------
          // Optimize graph and load the optimized graph onto a compute device
          // ------------------------------------------------------------------------
          
          // Optimize the network for a specific runtime compute device, e.g. CpuAcc, GpuAcc
          armnn::IRuntime::CreationOptions options;
          armnn::IRuntimePtr runtime(armnn::IRuntime::Create(options));
          armnn::OptimizerOptions optOptions;
          optOptions.m_shapeInferenceMethod == armnn::ShapeInferenceMethod::InferAndValidate;
          armnn::IOptimizedNetworkPtr optimizedNet = armnn::Optimize(*network,
																           programOptions.computeDevice,
																           runtime->GetDeviceSpec(),
																           optOptions);
          
          // Load the optimized network onto the runtime device
          armnn::NetworkId networkId;
          runtime->LoadNetwork(networkId, std::move(optimizedNet));

**Comment by fullm00n1** [comment]
I may not be understanding something correctly, but the call to Optimize with the InferAndValidate option seems to trigger optimize to call Graph::InterTensorInfo.  At the end of the loop in InferTensorInfos, it appears that the ValidateOnly section executes because in looks at the layer->m_ShapeInferenceMethod member of the layer which is set to ValidateOnly by default when the layer is created.  

Am I missing some step that would change the layer->m_ShapeInferenceMethod to something other than ValidateOnly?

**Comment by MikeJKelly** [comment]
Apologies, I asked around and there's a flag you need to set on the TFLiteParser too. 

```
     armnnTfLiteParser::ITfLiteParser::TfLiteParserOptions parserOptions;
     parserOptions.m_InferAndValidate = true;

     // Import the TensorFlowLite model.
     using IParser = armnnTfLiteParser::ITfLiteParser;
     auto armnnparser(IParser::Create(parserOptions));
```

Then on all created layers layer->m_ShapeInferenceMethod will be InferAndValidate. 

However, I can't be sure that we will be able to support your dynamic tensor shapes as our support for dynamic tensors is quite simple. Can you output your model with static shapes?

**Comment by fullm00n1** [comment]
Do you know how I can determine were the dynamic tensor shapes are being inserted?  We go from Matlab to ONXX to Tensorflow to Tensorflow lite.   It appears that the onnx-tf converter may be the one creating the dynamic shapes when going from ONNX to Tensorflow and you can't really control the conversion.  Do you have any suggestion on possibly a simple conversion within Tensorflow to convert from dynamic to static tensors?

**Comment by fullm00n1** [comment]
I was able to get Netron installed locally.  Is there a way within Netron to see if a tensor is dynamic or static?

**Comment by MikeJKelly** [comment]
Are you still getting an error when parsing the TFLite model? 

Dynamic tensors will have one or more -1 in their shape_signatures.

**Comment by TeresaARM** [comment]
Hi @fullm00n1,

I understand your problem is resolved. Please let us know otherwise.

**Comment by TeresaARM** [comment]
Hi @fullm00n1,

I am closing the issue as It seems your problem has been resolved. Please re-open the issue otherwise and thank you for reporting this

Kindest regards
